File: db/indexes.js
import { MongoClient } from "mongodb";

async function setupCollectionsAndIndexes(uri) {
  const client = new MongoClient(uri);

  try {
    await client.connect();
    const db = client.db("mileapp_db");

    await db.createCollection("users");
    console.log("created collection: users");

    await db.createCollection("tasks");
    console.log("created collection: tasks");

    const usersCollection = db.collection("users");

    await usersCollection.createIndex({ email: 1 }, { unique: true });
    console.log("created index on users.email (unique)");

    const tasksCollection = db.collection("tasks");

    await tasksCollection.createIndex({ status: 1 });
    console.log("created index on tasks.status");

    await tasksCollection.createIndex({ priority: 1 });
    console.log("created index on tasks.priority");

    await tasksCollection.createIndex({ due_date: 1 });
    console.log("created index on tasks.due_date");

    await tasksCollection.createIndex({ created_at: -1 });
    console.log("created index on tasks.created_at (descending)");

    console.log("collections and indexes setup complete.");
  } catch (err) {
    console.error("failed to setup collections or indexes:", err);
  } finally {
    await client.close();
  }
}

setupCollectionsAndIndexes("mongodb://localhost:27017");

File: pkg/database/indexes.go
package database

import (
	"context"
	"fmt"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

func SetupIndexes(ctx context.Context, db *mongo.Database) error {
	usersCollection := db.Collection("users")

	emailIndex := mongo.IndexModel{
		Keys:    bson.D{{Key: "email", Value: 1}},
		Options: options.Index().SetUnique(true),
	}

	if _, err := usersCollection.Indexes().CreateOne(ctx, emailIndex); err != nil {
		return fmt.Errorf("failed to create email index: %w", err)
	}

	tasksCollection := db.Collection("tasks")

	statusIndex := mongo.IndexModel{
		Keys: bson.D{{Key: "status", Value: 1}},
	}

	if _, err := tasksCollection.Indexes().CreateOne(ctx, statusIndex); err != nil {
		return fmt.Errorf("failed to create status index: %w", err)
	}

	priorityIndex := mongo.IndexModel{
		Keys: bson.D{{Key: "priority", Value: 1}},
	}

	if _, err := tasksCollection.Indexes().CreateOne(ctx, priorityIndex); err != nil {
		return fmt.Errorf("failed to create priority index: %w", err)
	}

	dueDateIndex := mongo.IndexModel{
		Keys: bson.D{{Key: "due_date", Value: 1}},
	}

	if _, err := tasksCollection.Indexes().CreateOne(ctx, dueDateIndex); err != nil {
		return fmt.Errorf("failed to create due_date index: %w", err)
	}

	createdAtIndex := mongo.IndexModel{
		Keys: bson.D{{Key: "created_at", Value: -1}},
	}

	if _, err := tasksCollection.Indexes().CreateOne(ctx, createdAtIndex); err != nil {
		return fmt.Errorf("failed to create created_at index: %w", err)
	}

	return nil
}
